app/controllers/discourse_ai/admin/ai_personas_controller.rb

```diff
@@ -74,8 +74,205 @@ def destroy
         end
       end
 
+      class << self
+        POOL_SIZE = 10
+        def thread_pool
+          @thread_pool ||=
+            Concurrent::CachedThreadPool.new(min_threads: 0, max_threads: POOL_SIZE, idletime: 30)
+        end
+
+        def schedule_block(&block)
+          # think about a better way to handle cross thread connections
+          if Rails.env.test?
+            block.call
+            return
+          end
+
+          db = RailsMultisite::ConnectionManagement.current_db
+          thread_pool.post do
+            begin
+              RailsMultisite::ConnectionManagement.with_connection(db) { block.call }
+            rescue StandardError => e
+              Discourse.warn_exception(e, message: "Discourse AI: Unable to stream reply")
+            end
+          end
+        end
+      end
+
+      CRLF = "\r\n"
+
+      def stream_reply
```

Comment:
I see Keegan mentioned a similar thing, but this is an extremely fat controller action, I really don't feel like we should be doing this very often, if ever. It doesn't feel very maintainable or easy to read, and could easily expand even further in size. Much of this is a perfect case for using a service.

Reply:
The reasons why it's perfect for a service:

* It has model lookups
* It has several parameter validations and error messages
* It has a clear action and result from that action, which could also have more errors

Not saying the reply streaming stuff should be in one, but everything before it easily could be

Reply:
I have been waiting a bit on service docs and service stability, stuff has been changing recently